### PR TITLE
Parameterize the `Security` plugin with the request-wise thread context supplier

### PR DESCRIPTION
**Pull request by albertzaharovits**
When `Security` authenticates a request (`AuthenticationService#authenticate`) it mostly looks in the thread context for all the request details that it needs.
Consequently, there's code that populates the thread context before running authentication.
This PR brings all the code that's responsible for populating the said thread context, as a parameter to the `Security` plugin.
This is required in the follow-up that is going to involve setting up the thread context (given the aforementioned parameter), and running the authentication under that context, before the request is dispatched (rather than "while" it's being dispatched, currently).

### CODE REVIEW

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java [hunk 1]
```diff
@@ -1625,15 +1631,29 @@ public boolean test(String profile, InetSocketAddress peerAddress) {
...
             } else {
                 sslConfiguration = null;
+                populateClientCertificate = (channel, threadContext) -> {};
             }
+            HttpServerTransport.Dispatcher dispatcherWithContext = HttpServerTransport.Dispatcher.dispatchWithThreadContextWrapper(
+                dispatcher,
+                (restRequest, threadContext) -> {
+                    dispatcherContext.accept(restRequest.getHttpRequest(), threadContext);
+                    populateClientCertificate.accept(restRequest.getHttpChannel(), threadContext);
+                    RemoteHostHeader.process(restRequest, threadContext);
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:1648]: This is the additional thread context that `Security` constructs, brought together in this single place. This context must be set, before `AuthenticationService#authenticate` is attempted.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java [hunk 2]
```diff
@@ -101,13 +96,7 @@ public void handleRequest(RestRequest request, RestChannel channel, NodeClient c
             return;
         }
 
-        if (extractClientCertificate) {
-            HttpChannel httpChannel = request.getHttpChannel();
-            SSLEngineUtils.extractClientCertificates(logger, threadContext, httpChannel);
-        }
-
         final RestRequest wrappedRequest = maybeWrapRestRequest(request);
-        RemoteHostHeader.process(request, threadContext);
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java:110]: These things moved.

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Plugin.java [hunk 3]
```diff
@@ -99,17 +102,22 @@ public Map<String, Supplier<HttpServerTransport>> getHttpTransports(
...
         NetworkService networkService,
         HttpServerTransport.Dispatcher dispatcher,
+        BiConsumer<HttpPreRequest, ThreadContext> dispatcherContext,
         ClusterSettings clusterSettings,
         Tracer tracer
     ) {
+        HttpServerTransport.Dispatcher dispatcherWithContext = HttpServerTransport.Dispatcher.dispatchWithThreadContextWrapper(
+            dispatcher,
+            (httpPreRequest -> dispatcherContext.accept(httpPreRequest, threadPool.getThreadContext()))
+        );
```
Comment (albertzaharovits) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Plugin.java:112]: Previously, the `RestController` was in charge to build the thread context from the request headers.
Now, the `dispatcherContext` does that, and it is exposed by the `ActionModule`, to be used by the plugins that build `HttpServerTransport` implementations (which take the `RestController` as the argument).

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java [hunk 4]
```diff
@@ -1592,6 +1596,7 @@ public Map<String, Supplier<HttpServerTransport>> getHttpTransports(
         NamedXContentRegistry xContentRegistry,
         NetworkService networkService,
         HttpServerTransport.Dispatcher dispatcher,
+        BiConsumer<HttpPreRequest, ThreadContext> dispatcherContext,
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:1599]: This is important. The `Security` plugin needs access to this supplier of thread context.

**Review by jakelandis — COMMENTED** [review]
change is looking good, thanks for the change away from the dispatch wrapper i think this reads much better.  just a couple minor questions and nitpicks

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Plugin.java [hunk 5]
```diff
@@ -99,6 +103,7 @@ public Map<String, Supplier<HttpServerTransport>> getHttpTransports(
         NamedXContentRegistry xContentRegistry,
         NetworkService networkService,
         HttpServerTransport.Dispatcher dispatcher,
+        BiConsumer<HttpPreRequest, ThreadContext> setDispatchContext,
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Plugin.java]: Pushed [ff9e3ca](https://github.com/elastic/elasticsearch/pull/95040/commits/ff9e3ca53635aeaa3b9249e977343b01be924c96) .

File: server/src/main/java/org/elasticsearch/action/ActionModule.java [hunk 6]
```diff
@@ -539,19 +545,49 @@ public ActionModule(
...
+     * Certain request header values need to be copied in the thread context under which request handlers are to be dispatched.
+     * Careful that this method modifies the thread context. The thread context must be reinstated after the request handler
+     * finishes and returns.
+     */
+    public void copyRequestHeadersToThreadContext(HttpPreRequest request, ThreadContext threadContext) {
+        copyRequestHeadersToThreadContext(request, headersToCopy, threadContext);
+    }
+
+    // to be used in tests
+    public static void copyRequestHeadersToThreadContext(
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply server/src/main/java/org/elasticsearch/action/ActionModule.java]: Pushed [a2c7b74](https://github.com/elastic/elasticsearch/pull/95040/commits/a2c7b746e97e4dd6b0eb76c1b71f85b977167047) .

File: server/src/test/java/org/elasticsearch/http/AbstractHttpServerTransportTests.java [hunk 7]
```diff
@@ -198,6 +204,170 @@ public HttpStats stats() {
...
+
+        };
+        // the set of headers to copy
+        final Set<RestHeaderDefinition> headers = new HashSet<>(
+            Arrays.asList(new RestHeaderDefinition("header.1", true), new RestHeaderDefinition("header.2", true))
+        );
+        final Map<String, List<String>> restHeaders = new HashMap<>();
+        restHeaders.put("header.1", Collections.singletonList("true"));
+        restHeaders.put("header.2", Collections.singletonList("true"));
+        restHeaders.put("header.3", Collections.singletonList("false"));
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply server/src/test/java/org/elasticsearch/http/AbstractHttpServerTransportTests.java]: Sure! I'm sorry for the wasted time. Pushed [757dcb6](https://github.com/elastic/elasticsearch/pull/95040/commits/757dcb6abe8db1f3bb1e6aede2543f349fc088b7) .

File: server/src/main/java/org/elasticsearch/rest/RestController.java [hunk 8]
```diff
@@ -69,9 +68,9 @@ public class RestController implements HttpServerTransport.Dispatcher {
      */
     static final Set<String> SAFELISTED_MEDIA_TYPES = Set.of("application/x-www-form-urlencoded", "multipart/form-data", "text/plain");
 
-    static final String ELASTIC_PRODUCT_HTTP_HEADER = "X-elastic-product";
+    public static final String ELASTIC_PRODUCT_HTTP_HEADER = "X-elastic-product";
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply server/src/main/java/org/elasticsearch/rest/RestController.java]: Not anymore, good catch! Pushed [91f372b](https://github.com/elastic/elasticsearch/pull/95040/commits/91f372b8d8d89cc15af506e6019b86e3f1736b01)

File: server/src/test/java/org/elasticsearch/http/AbstractHttpServerTransportTests.java [hunk 9]
```diff
@@ -198,6 +204,170 @@ public HttpStats stats() {
...
+                }
+
+                @Override
+                protected void populateRequestThreadContext(RestRequest restRequest, ThreadContext threadContext) {
+                    ActionModule.copyRequestHeadersToThreadContext(restRequest.getHttpRequest(), headers, threadContext);
+                }
+            }
+        ) {
+            transport.dispatchRequest(fakeRequest, channel, null);
+            assertNull(threadPool.getThreadContext().getHeader("header.1"));
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply server/src/test/java/org/elasticsearch/http/AbstractHttpServerTransportTests.java:291]: Correct, I commented as such in the code: [3e69804](https://github.com/elastic/elasticsearch/pull/95040/commits/3e698046e2895968ec8cb943e2517e96a15cd7eb)

File: server/src/test/java/org/elasticsearch/http/AbstractHttpServerTransportTests.java [hunk 10]
```diff
@@ -198,6 +204,170 @@ public HttpStats stats() {
...
+                }
+
+                @Override
+                protected void populateRequestThreadContext(RestRequest restRequest, ThreadContext threadContext) {
+                    ActionModule.copyRequestHeadersToThreadContext(restRequest.getHttpRequest(), headers, threadContext);
+                }
+            }
+        ) {
+            transport.dispatchRequest(fakeRequest, channel, null);
+            assertThat(threadPool.getThreadContext().getHeader(Task.TRACE_ID), nullValue());
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply server/src/test/java/org/elasticsearch/http/AbstractHttpServerTransportTests.java:374]: Same as https://github.com/elastic/elasticsearch/pull/95040#discussion_r1161551098 .

**Comment by albertzaharovits** [comment]
Thank you for the speedy reviews @jakelandis !
Please take another look.

**Review by jakelandis — APPROVED** [review]
LGTM, thanks for the discussion and updates

**Pull request merged**